pkg/chaosctl/physicalmachine/ssh.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package physicalmachine
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+type SshTunnel struct {
+	config *ssh.ClientConfig
+	host   string
+	port   string
+	client *ssh.Client
+}
+
+func NewSshTunnel(ip, port string, user, privateKeyFile string) (*SshTunnel, error) {
+	hostKey, err := getHostKey(ip)
+	if err != nil {
+		return nil, err
+	}
+	config := ssh.ClientConfig{
+		Timeout: 5 * time.Minute,
+		User:    user,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeysCallback(func() ([]ssh.Signer, error) {
+				key, err := ioutil.ReadFile(privateKeyFile)
+				if err != nil {
+					return nil, fmt.Errorf("ssh key file read failed: %v", err)
+				}
+
+				signer, err := ssh.ParsePrivateKey(key)
+				if err != nil {
+					return nil, fmt.Errorf("ssh key signer failed: %v", err)
+				}
+				return []ssh.Signer{signer}, nil
+			}),
+		},
+		HostKeyCallback: ssh.FixedHostKey(hostKey),
+	}
+	return &SshTunnel{
+		config: &config,
+		host:   ip,
+		port:   port,
+	}, nil
+}
+
+func (s *SshTunnel) Open() error {
+	// first, try ssh with public keys
+	conn, err := ssh.Dial("tcp", net.JoinHostPort(s.host, s.port), s.config)
+	if err != nil {
+		// if failed, try ssh with password
```

Comment:
how about printing the error message?

Reply:
As https://github.com/chaos-mesh/chaos-mesh/pull/2642#discussion_r775411415 suggested, I have already removed this retry logic :)